llvm/docs/DeveloperPolicy.rst

```diff
@@ -853,6 +853,69 @@ their patch with every possible configuration.
+and command line arguments) both as part of the issue description and as a
+godbolt.org link. An effort should be made to deduplicate issues that likely
+have the same root cause, and check whether a similar issue has already been
+reported. Reports should always be submitted against current LLVM ``main``,
+not a released version.
+
+If possible, provide information on when an issue was introduced (e.g. by
+checking older versions on godbolt). A regression from the last LLVM release
+has higher priority than an issue that existed for decades.
+
+The remaining guidelines depend on the type of issue the fuzzer detects.
+
+**For miscompilations:** These issues are usually detected by looking for
+different results when using ``-O0`` and ``-O2``, or similar. When reporting
+miscompilations, please make sure that your fuzzing methodology can only
+generate well-defined, deterministic code. Results between optimizations levels
+can legitimately differ if the code invokes undefined behavior, or includes
+non-deterministic operations. Note that running cleanly under sanitizers is
+not sufficient to establish absense of undefined behavior.
+
+Reports using ``-Ofast``, ``-ffast-math``, or other flags that permit
+floating-point reassociation/approximation must include a credible root cause
+analysis, as behavior differences are likely to be caused by legal transforms.
+
+**For crashes / assertion failures:** Crashes that occur on valid code are more
+valuable than crashes on invalid code. Both can be reported, but the former is
+more likely to see a timely fix.
+
+Fuzzing can be performed at multiple levels, where higher levels are less likely
+to produce false positives. For example, a crash triggered by valid C code will
+generally indicate a real bug. However, a crash triggered by syntactically
+well-formed LLVM IR may not. For example, a target that does not support
+scalable vectors may break when provided IR using them. When fuzzing at a lower
+level, it is encouraged to verify the plausibility of the results.
+
+When fuzzing LLVM IR, fatal errors that do not generate a stack trace should
+not be reported. They indicate an incorrect use of LLVM, rather than a bug.
+
+**For missed optimizations:** There is an infinite number of optimizations that
+*could* be implemented, but only a small subset of them is relevant for
+real-world code. As such, fuzzer-generated reports for missed optimizations are
+only accepted if plausible real-world usefulness can be shown.
+
+For example, a valid strategy is to take a corpus of real-world code and use a
+super-optimizer to find missed optimization opportunities. An invalid strategy
+is to generate random code and check whether GCC generates less code than Clang.
+
+Fuzzer-generated missed optimization reports that are not derived from
+real-world code must include a root-cause analysis, and an explanation for why
+you believe that the missed optimization has real-world relevance.
+
```

Comment:
Should add some comments on non-deterministic bugs (compiler should be run with ASLR disabled).
1. IR diff
2. textual IR diff (e.g., instruction order/incoming BB order)
3. statistics diff

I think it is important for reproducible builds.

Reply:
I don't think ASLR enabled/disabled has anything to do with it. If the crash/miscompilation is non-deterministically reproducible, that will happen regardless of text / data / etc sections of the compiler being located at a different memory location each time (I would agree if we were debugging a kernel bug). The addresses in the stack trace will likely be different anyways (due to, e.g., different env variables). What should probably matter is the full clang command (target/cpu/arch included).

Reply:
>  If the crash/miscompilation is non-deterministically reproducible

Perhaps you are mistaken. There is no crash/miscompilation. I mean compiler should produce the same result under the same environment settings.

> I don't think ASLR enabled/disabled has anything to do with it.

Agree. Compiler should produce deterministic result regardless of non-determinism in memory allocation.

Reply:
> Perhaps you are mistaken. There is no crash/miscompilation. I mean compiler should produce the same result under the same environment settings.

I see, I thought you were referring to miscompilations/crash that happen non-deterministically (like, e.g., some we saw in GVNSink, IIRC).

Reply:
I don't really get what the suggested change to the text here is. Do you want to add a fourth category next to miscompile/crash/missed optimization?

Reply:
> I don't really get what the suggested change to the text here is. Do you want to add a fourth category next to miscompile/crash/missed optimization?

Yeah.